### DISCUSSION TITLE
Typesafety:   Weird behaviour today

### DISCUSSION THREAD

**Discussion by OsirisTerje** [nunit/nunit#4904] [Problems]
The code below works, it compiles, the tests are green, and no analyzer warnings.
The code doesn't make sense - except only if you have some very weird way of thinking about testing.

Should we work through the different constraints and make sure they "make sense" with respect to the types they are intended for ?

![image](https://github.com/user-attachments/assets/86590a18-986e-4a67-8c43-cd9ba4ff7a67)

@manfred-brands @stevenaw @mikkelbu @jnm2 @CharliePoole 

PS:  The last one is quite fun, the length doesn't matter, but it actually seems to compare them alphabetically.

**Comment by jnm2** [comment]
The runtime declares `false.CompareTo(true)` and `"Whatever".CompareTo("Hello")`, so these don't strike me as nonsensical. The types in question have their own well-known definition of being greater or less.

There's a bit of an issue around CompareTo with types that aren't IComparable. In this case, .NET itself is a lot like NUnit. There's no type safety. If you call `Nullable.CompareTo` or `Comparer<T>.Default.Compare`, you may hit:

> System.ArgumentException: 'At least one object must implement IComparable.'

The same is also true of `(new object(), new object()).CompareTo((new object(), new object()));` or `new ValueTuple<object>(new()).CompareTo(new ValueTuple<object>(new()));`.

**Comment by OsirisTerje** [comment]
Good point.
I see from this:  https://learn.microsoft.com/en-us/dotnet/api/system.icomparable.compareto?view=net-9.0
that

> Compares the current instance with another object of the same type and returns an integer that indicates whether the current instance precedes, follows, or occurs in the same position in the sort order as the other object.

So our GreaterThan with strings will be dependent upon the sort order.  

Still think GreaterThan looks silly in these cases though,  but I rest my case :-)

**Reply by jnm2** [reply]
IComparable is the basis for sorting but also just determining `<` and `>` much of the time. For instance, `Max` or `Min` functions pick the greatest or least using IComparable, and IsBetween/Clamp functions would operate on the same principle, all without thinking of IComparable as being about sort order. It can be used that way, but it's really more about what is greater or lesser of two objects, however the objects define it.

I see how it could look unusual if "greater than" doesn't feel like a synonym for ordering! Partly, this brings up a possible design philosophy. If programmers don't make themselves sufficiently clear in their coding, that's not the API or language's fault for being too flexible, if there was another choice.

**Reply by stevenaw** [reply]
It is an interesting curiousity that "max" < "min" in this case 😀

**Reply by jnm2** [reply]
In which case?

**Reply by manfred-brands** [reply]
Alphabetically

**Reply by jnm2** [reply]
Ah nice 😆

**Comment by CharliePoole** [comment]
@OsirisTerje Interesting observation. It seems as if GreaterThan used with sorted strings is an artifact of the programming profession. In normal speech, one says "follows" or "comes after" or at worst "sorts after." :-)

**Comment by manfred-brands** [comment]
@OsirisTerje I see you already closed it, but I agree with @jnm2. It is fine for testing expected sort order.
